tests/kafkatest/tests/streams/streams_cooperative_rebalance_upgrade_test.py

```diff
@@ -0,0 +1,191 @@
+
+        # now verify tasks are unique
+        for processor in processors:
+            self.get_tasks_for_processor(processor)
+            self.logger.info("Active tasks %s" % processor.active_tasks)
+
+        overlapping_tasks = processor1.active_tasks.intersection(processor2.active_tasks, processor3.active_tasks)
+        assert len(overlapping_tasks) == int(0), \
+            "Final task assignments are not unique %s %s %s" % (processor1.active_tasks, processor2.active_tasks, processor3.active_tasks)
+
+        other_overlapping_tasks = processor2.active_tasks.intersection(processor3.active_tasks)
+        assert len(other_overlapping_tasks) == int(0), \
+            "Final task assignments are not unique %s %s" % (processor2.active_tasks, processor3.active_tasks)
+
+        # test done close all down
+        stop_processors(processors, self.second_bounce_phase + self.stopped_message)
+
+        self.producer.stop()
+        self.kafka.stop()
+        self.zookeeper.stop()
+
+    def maybe_upgrade_rolling_bounce_and_verify(self,
+                                                processors,
+                                                previous_phase,
+                                                current_phase,
+                                                upgrade_from_version=None):
+        for processor in processors:
+            # stop the processor in prep for setting "update.from" or removing "update.from"
+            verify_stopped(processor, previous_phase + self.stopped_message)
+            # upgrade to version with cooperative rebalance
+            processor.set_version("")
+            processor.set_upgrade_phase(current_phase)
+
+            if upgrade_from_version is not None:
+                # need to remove minor version numbers for check of valid upgrade from numbers
+                upgrade_version = upgrade_from_version[:upgrade_from_version.rfind('.')]
+                rebalance_mode_msg = self.cooperative_turned_off_msg % upgrade_version
+            else:
+                upgrade_version = None
+                rebalance_mode_msg = self.cooperative_enabled_msg
+
+            self.set_props(processor, upgrade_version)
+            node = processor.node
+            with node.account.monitor_log(processor.STDOUT_FILE) as stdout_monitor:
+                with node.account.monitor_log(processor.LOG_FILE) as log_monitor:
+                    processor.start()
+                    # verify correct rebalance mode either turned off for upgrade or enabled after upgrade
+                    log_monitor.wait_until(rebalance_mode_msg,
+                                           timeout_sec=60,
+                                           err_msg="Never saw '%s' message " % rebalance_mode_msg + str(processor.node.account))
+
+                # verify rebalanced into a running state
+                rebalance_msg = current_phase + self.running_state_msg
+                stdout_monitor.wait_until(rebalance_msg,
```

Comment:
Hmm, on old versions, we do not have state listener and hence would not print `current_phase + self.running_state_msg`, how we would still pass here?

Reply:
At this point in the test, we've upgraded to the new version so we have a state listener.